### PR TITLE
lib: add `data_size` and `byte_unit`

### PR DESCRIPTION
**Pull request by simonvonhackewitz**
Example
```
❯ fz -e 'x := units.data_size.tb 1; say "A {x.as_string_si} HDD is only $x."' 
A 1.00 TB HDD is only 931 GiB.
```

Needs some testing, especially edge cases. ~~Everything above ≈15 EiB (or ≈18 EB) currently causes errors due to the maximum value of u64.~~
Maximum representable value with `u128` is 340,282,366 QB or 268,435,455 QiB.

@michaellilltokiwa I'd like some feedback on the overall approach and how it is structured.

### CODE REVIEW

**Comment by michaellilltokiwa** [comment]
I think it might be easier to have just one type for data_size, private constructor and two inititing features, if necessary:
`data_size.type.from_si arg`,  `data_size.type.from_iec arg`

Does that make sense?

**Comment by simonvonhackewitz** [comment]
Is there a convention for how many decimal places should be displayed? And when should the next larger unit be used?
Also, some control on when to add leading spaces for alignment would be good. They are useful when printing a table, but it looks strange in a single sentence.

**Comment by michaellilltokiwa** [comment]
@simonvonhackewitz I would do it similar to:
```
  # create a string representation of this duration using the given unit.
  #
  public as_string(u time.time_unit) String =>
    n := $(nanos / u.in_ns)
    " "*(max 0 4-n.byte_length) + n + u.short_name
```

File: modules/base/src/units.fz [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#
+# You should have received a copy of the GNU General Public License along with The
+# Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------------
+#
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion standard library feature data_size
```
Comment (michaellilltokiwa) [inline modules/base/src/units.fz]:
```suggestion
#  Source code of Fuzion standard library feature units
```

File: modules/base/src/units.fz [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#
+# -----------------------------------------------------------------------
+
+# NYI comment
+#
```
Comment (michaellilltokiwa) [inline modules/base/src/units.fz]:
```suggestion
# units -- unit feature to group unit related features like data_size
#
```

**Review by michaellilltokiwa — CHANGES_REQUESTED** [review]
Good but please add a comment what iec/si means. Also either add a test or an issue that tests for this are still missing.

**Comment by michaellilltokiwa** [comment]
@simonvonhackewitz Hmm yes.
A good test is almost like a piece of art. In this particular case I see two goals of a test:
- test edge cases (e.g. u128.zero, u128.max) to gain confidence in the algorithm of e.g. as_string
- making sure code stays working as intended. The intention is either encoded in the test or the recording.

The test you added is good but I would probably focus more on edge cases than generate some more or less _random_ numbers.

File: modules/base/src/units/data_size.fz [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+  public fixed infix - (other data_size) data_size
+  pre
+    safety: data_size.this >= other
+  =>
+    data_size (bytes - other.bytes)
+
+
+  # this units.data_size multiplied by factor n
+  #
+  public infix * (n u128) data_size
```
Comment (michaellilltokiwa) [inline modules/base/src/units/data_size.fz]: I missed this in my last review.
What is the rationale of `infix *` and `infix /`? 
I don't see that these make much sense?

**Review by michaellilltokiwa — APPROVED** [review]
Great!
